README.md
# Enlighten

Check it out at https://enlighten-game.herokuapp.com/

Enlighten is a collaborative webgame. You are a little square, navigating a dark maze with only a small torch.

Use WASD or Arrow keys to move around the maze. If you collide with the walls of the maze, you leave a small light showing where you died, and return to the start of the maze. Find the way out by learning from your and others' mistakes.

To help, there are campfires: permanent light sources scattered across the maze. Walk over them to refuel your decaying torch. You can see other players' lights, and they can see yours too. The more players online, the easier it is to complete a level.


CHANGELOG.md
# Changelog

### v2.1.0
- Smooth scrolling
- Added player lights ("torches")
- Added permanent light sources ("campfires")
- Added basic instructions page

### v2.0.1
- Add two new levels
- Finish position is now rectangular

### v2.0.0
- Multiple levels added
  - Respawn logic updated
  - Each room has
    - Array of lights
    - Only shows players in it
    - Start / finish positions
- Basic start button added
- Displays number of players online

### v1.3.1
- Enable gzip compression
- socket.io CDN used

### v1.3.0
- New website: [Heroku Deploy](https://enlighten-game.herokuapp.com)
  - Lower ping, higher fps
- Remove Room Creator temporarily

### v1.2.0
- Completely remove room sending
  - Room only sent when client connects to server
  - Room drawn once to buffer, buffer drawn only when view offset is changed

### v1.1.1
- Remove Mouse/Player positions
- Save full light gradient for faster drawing
- Create buffer for `drawRoom()`
- Only emit `playerMove` when player moves (instead of `gameLoop()`)

### v1.1.0
- Fix crashing when player dies/disconnects
- Clean up `movePlayers()` using clamp
- More accurate light finding in `newLight()`
- Created simple room creator
- New starting room

### v1.0.0
- Initial upload
- Available at [nodejs-warblewarble.rhcloud.com](nodejs-warblewarble.rhcloud.com) (prone to change)


client/main.js
//Top canvas: shadow that covers game
var ct = document.getElementById("cantop");
var ctx = ct.getContext("2d");
//Bot canvas: visible portion of room
var cb = document.getElementById("canbot");
var cbx = cb.getContext("2d");
ct.width = cb.width = window.innerWidth;
ct.height = cb.height = window.innerHeight;
ct.focus();

var config = {
  "tileSize": 50,  //side length of tile (tiles are square)
  "radius": 100,   //radius of light
  "pradius": 200, //radius of player light
  "border": 300,   //border where scrolling begins
  "playerSize": 24 //side length of player (player is square)
};

var socket; //socket connection with server
var lights = []; //lights
var campfires = []; //campfires

var roomBuffer = document.createElement("canvas"); //canvas of entire room
var rbx = roomBuffer.getContext("2d");
var roomDraw = true; //redraw room (due to changing view)

//player object
var player = {
  x: 0,        //position coords
  y: 0,
  xoffset: 0,  //offset for view (top-left coords)
  yoffset: 0,
  light: 1.0
};
var others = [];
var move = {
  up: false,
  down: false,
  left: false,
  right: false
};

var animloopHandle;
var oldtime = new Date().getTime(), time = new Date().getTime(); //for fps

var lightGradient = ctx.createRadialGradient(0,0,0,0,0,config.radius); //light gradient
lightGradient.addColorStop(0,"white");
lightGradient.addColorStop(1,"transparent");

var debug = false;

window.onload = function(){
  var startBtn = document.getElementById("startbtn");
  startBtn.onclick = function(){
    startGame();
  };

  if(debug){
    ct.addEventListener("mousedown", function(event){
      socket.emit("addLight", {x: event.pageX + player.xoffset, y: event.pageY + player.yoffset});
    });
  }

  ct.addEventListener("keydown", function(event){ //press key down
    var key = event.which || event.keyCode;
    if(key === 87 || key == 38) move.up = true;
    if(key === 83 || key === 40) move.down = true;
    if(key === 65 || key === 37) move.left = true;
    if(key === 68 || key === 39) move.right = true;
    socket.emit("playerMove", moveEncode());
  });

  ct.addEventListener("keyup", function(event){ //depress key
    var key = event.which || event.keyCode;
    if(key === 87 || key == 38) move.up = false;
    if(key === 83 || key === 40) move.down = false;
    if(key === 65 || key === 37) move.left = false;
    if(key === 68 || key === 39) move.right = false;
    socket.emit("playerMove", moveEncode());
  });
};

window.requestAnimFrame = (function(){
  return  window.requestAnimationFrame       ||
          window.webkitRequestAnimationFrame ||
          window.mozRequestAnimationFrame    ||
          function(callback){
            window.setTimeout(callback, 1000 / 60);
          };
})();

window.cancelAnimFrame = (function(handle) {
    return  window.cancelAnimationFrame       ||
            window.webkitCancelAnimationFrame ||
            window.mozCancelAnimationFrame;
})();

function startGame(){
  if(!socket){
    socket = io();
    setupSocket();
  }
  if(!animloopHandle) animloop();
  socket.emit("startClient");
}

function moveEncode(){ //enocde move into 4-bit
  var result = 0;
  if(move.up) result += 1;
  if(move.down) result += 2;
  if(move.left) result += 4;
  if(move.right) result += 8;
  return result;
}

function setupSocket(){
  socket.on("startServer", function(newPlayer, otherPlayers, room, newCampfires){
    player = newPlayer;
    document.getElementById("room").innerHTML = "Room: " + player.room;
    player.sWidth = window.innerWidth;
    player.sHeight = window.innerHeight;
    others = otherPlayers;

    updateRoom(room);

    campfires = newCampfires;

    document.getElementById("menu").style.display = "none";
    ct.focus();

    socket.emit("confirm", player);
  });

  socket.on("newLights", function(newLights){
    lights = newLights;
  });

  socket.on("newPosition", function(newPlayer, newOthers, numPlayers){
    player.x = newPlayer.x;
    player.y = newPlayer.y;
    player.light = newPlayer.light;
    if(player.xoffset !== newPlayer.xoffset || player.yoffset !== newPlayer.yoffset) roomDraw = true;
    player.xoffset = newPlayer.xoffset;
    player.yoffset = newPlayer.yoffset;
    others = newOthers;

    document.getElementById("numPlayer").innerHTML = "Players: " + numPlayers;
  });

  socket.on("newRoom", function(newPlayer, newRoom, newCampfires){
    player = newPlayer;
    document.getElementById("room").innerHTML = "Room: " + player.room;
    updateRoom(newRoom);
    campfires = newCampfires;
  });

  setInterval(function(){
    document.getElementById("fps").innerHTML = "FPS: "+Math.round(1000/(time-oldtime));
    socket.emit("ping", new Date().getTime());
  }, 1000);

  socket.on("pong", function(date){
    document.getElementById("ping").innerHTML = "Ping: "+ (new Date().getTime() - date);
  });

  socket.on("dead", function(newPlayer){
    player = newPlayer;
    window.setTimeout(function(){
      /*if(animloopHandle){
          window.cancelAnimationFrame(animloopHandle);
          animloopHandle = undefined;
      }*/
      //startGame();
      socket.emit("respawn");
    }, 1000);
  });
}

function updateRoom(room){
  roomBuffer.width = room[0].length*config.tileSize;
  roomBuffer.height = room.length*config.tileSize;
  for(var i=0;i<room.length;i++){
    for(var j=0;j<room[i].length;j++){
      if(room[i][j] === "1") rbx.fillStyle = "#ff0000";
      else rbx.fillStyle = "#00ff00";
      rbx.fillRect(j*config.tileSize, i*config.tileSize, config.tileSize, config.tileSize);
    }
  }
  roomDraw = true;
}

function animloop(){
  animloopHandle = requestAnimFrame(animloop);
  gameLoop();
}

function gameLoop(){
  if(roomDraw){
    drawRoom();
    roomDraw = false;
  }

  ctx.fillStyle = "#000000";
  ctx.fillRect(0, 0, player.sWidth, player.sHeight);
  drawLights();
  drawOthers();
  if(!player.dead) drawPlayer();

  oldtime = time;
  time = new Date().getTime();
}

function drawRoom(){
  cbx.drawImage(roomBuffer, -player.xoffset, -player.yoffset);
  //cbx.drawImage(roomBuffer, player.xoffset, player.yoffset, player.sWidth, player.sHeight, 0, 0, player.sWidth, player.sHeight);
}

function drawLights(){
  ctx.globalCompositeOperation = "destination-out";
  if(debug){ //make maze visible
    ctx.fillStyle = "rgba(0,0,0,1)";
    ctx.fillRect(0, 0, player.sWidth, player.sHeight);
  }

  ctx.fillStyle = lightGradient;
  for(var i=0;i<lights.length;i++){
    if(lights[i].fade === 1.0){ //full sized light
      ctx.translate(lights[i].x-player.xoffset, lights[i].y-player.yoffset);
      ctx.fillRect(-config.radius, -config.radius, 2*config.radius, 2*config.radius);
      ctx.translate(-lights[i].x+player.xoffset, -lights[i].y+player.yoffset);
    }
    else{ //waning light
      var grd = ctx.createRadialGradient(lights[i].x-player.xoffset,lights[i].y-player.yoffset,0,lights[i].x-player.xoffset,lights[i].y-player.yoffset,config.radius);
      grd.addColorStop(0,"black");
      grd.addColorStop(lights[i].fade,"transparent");
      ctx.fillStyle = grd;
      ctx.fillRect(lights[i].x-config.radius-player.xoffset,lights[i].y-config.radius-player.yoffset,2*config.radius,2*config.radius);
      ctx.fillStyle = lightGradient;
    }
  }

  for(var j=0;j<campfires.length;j++){
    drawSingleLight(campfires[j].x * config.tileSize - player.xoffset, campfires[j].y * config.tileSize - player.yoffset, config.pradius, 1);
  }

  //draw player lights
  drawSingleLight(player.x - player.xoffset, player.y - player.yoffset, config.pradius, player.light);

  //draw enemy lights
  for(var k=0;k<others.length;k++){
    drawSingleLight(others[k].x - player.xoffset, others[k].y - player.yoffset, config.pradius, others[k].light);
  }

  ctx.globalCompositeOperation = "source-over";
}

function drawSingleLight(x, y, radius, dim){
  var lightStyle = ctx.createRadialGradient(x, y, 0, x, y, radius);
  lightStyle.addColorStop(0, "black");
  lightStyle.addColorStop(dim, "transparent");
  ctx.fillStyle = lightStyle;
  ctx.fillRect(x - radius, y - radius, 2 * radius, 2 * radius);
  ctx.fillStyle = lightGradient;
}

function drawOthers(){
  for(var i=0;i<others.length;i++){
    ctx.fillStyle = "hsl(" + others[i].hue + ", 70%, 50%)";
    ctx.fillRect(others[i].x-config.playerSize/2-player.xoffset, others[i].y-config.playerSize/2-player.yoffset, config.playerSize, config.playerSize);
  }
}

function drawPlayer(){
  //ctx.strokeStyle = 'hsl(' + player.hue + ', 80%, 40%)';
  ctx.fillStyle = "hsl(" + player.hue + ", 70%, 50%)";
  ctx.fillRect(player.x-config.playerSize/2-player.xoffset, player.y-config.playerSize/2-player.yoffset, config.playerSize, config.playerSize);
}

window.addEventListener('resize', function() {
  player.sWidth = ct.width = cb.width = window.innerWidth;
  player.sHeight = ct.height = cb.height = window.innerHeight;
  roomDraw = true;
  socket.emit("resize", player.sWidth, player.sHeight);
});


server/index.js
var express = require("express");
var app     = express();
var comp    = require("compression");
var http    = require("http").createServer(app);
var io      = require("socket.io").listen(http);
var rooms   = require("./room.json");

app.use(comp());
app.use(express.static(__dirname + '/../client'));
var port = process.env.PORT  || 5000;
http.listen(port, function(){
  console.log("listening on:" + port);
});

var config = {
  "tileSize": 50,
  "radius": 100,
  "border": 300,
  "playerSize": 24
};
var speed = 0.2;
var lightLength = 5*60*1000; //5 minutes
var plightDecay = 0.0005;

var sockets = {};
var players = [];
var lights = [];
var campfires = [];
for(var i=0;i<rooms.length;i++){
  lights.push([]);
  campfires.push([]);
  for(var j=0;j<rooms[i].campfire.length;j++){
    campfires[i].push(rooms[i].campfire[j]);
  }
}

io.on("connection", function(socket){
  var currPlayer = {
    id: socket.id,
    x: rooms[0].startx * config.tileSize,
    y: rooms[0].starty * config.tileSize,
    xoffset: 0,
    yoffset: 0,
    hue: Math.round(Math.random() * 360),
    light: 1.0,
    move:{
      up: false,
      down: false,
      left: false,
      right: false,
    },
    room: 0,
    roomWidth: rooms[0].grid[0].length * config.tileSize,
    roomHeight: rooms[0].grid.length * config.tileSize,
    dead: false
  };

  socket.on("startClient", function(){
    var index = findID(currPlayer.id);
    if(index > -1) players.splice(index, 1);
    socket.emit("startServer", currPlayer, players, rooms[0].grid, campfires[0]);
  });

  socket.on("confirm", function(data){
    currPlayer = data;
    currPlayer.x = rooms[currPlayer.room].startx * config.tileSize;
    currPlayer.y = rooms[currPlayer.room].starty * config.tileSize;
    currPlayer.move = {
      up: false,
      down: false,
      left: false,
      right: false
    };
    players.push(currPlayer);
    sockets[currPlayer.id] = socket;
    console.log("New Player: " + currPlayer.id);
  });

  socket.on("respawn", function(){
    currPlayer.x = rooms[currPlayer.room].startx * config.tileSize;
    currPlayer.y = rooms[currPlayer.room].starty * config.tileSize;
    currPlayer.move = {
      up: false,
      down: false,
      left: false,
      right: false
    };
    currPlayer.dead = false;
    socket.emit("startServer", currPlayer, players, rooms[currPlayer.room].grid, campfires[currPlayer.room]);
  });

  socket.on("addLight", function(data){ //only used in debug
    newLight(data.x, data.y, currPlayer.room);
  });

  socket.on("playerMove", function(data){
    currPlayer.move.up = data & 1;
    currPlayer.move.down = data & 2;
    currPlayer.move.left = data & 4;
    currPlayer.move.right = data & 8;
  });

  socket.on("resize", function(sWidth, sHeight){
    currPlayer.sWidth = sWidth;
    currPlayer.sHeight = sHeight;
  });

  socket.on("ping", function(date){
    socket.emit("pong", date);
  });

  socket.on('disconnect', function(){
    //sockets[currPlayer.id] = undefined;
    var index = findID(currPlayer.id);
    if(index > -1) players.splice(index, 1);
    console.log("Player dc: " + currPlayer.id);
  });
});

function clamp(num, min, max){
  return Math.min(Math.max(num, min), max);
}

function findID(id){
  for(var i=0;i<players.length;i++){
    if(players[i].id === id) return i;
  }
  return -1;
}

function findLight(id, room){
  for(var i=0;i<lights[room].length;i++){
    if(id === lights[room].id) return i;
  }
  return -1;
}

function newLight(xx, yy, room){
  var newid = (xx + yy) * (xx + yy + 1) / 2 + yy;
  lights[room].push({x: xx, y: yy, id: newid, fade: 1.0});
  setTimeout(function(){
    var int = setInterval(function(){
      var index = findLight(newid, room);
      if(index > -1){
        lights[room][index].fade -= 0.02;
        if(lights[room][index].fade <= 0){
          lights[room].splice(index, 1);
          clearInterval(int);
        }
      }
    }, 20);
  }, lightLength);
}

/*function find(array, element, property){
  for(var i=0;i<array.length;i++){
    if(array[i][property] === element) return i;
  }
  return -1;
}*/

function loadNextLevel(player){
  player.room++;
  if(player.room >= rooms.length) player.room = rooms.length - 1;
  player.x = rooms[player.room].startx * config.tileSize;
  player.y = rooms[player.room].starty * config.tileSize;
  player.roomWidth = rooms[player.room].grid[0].length * config.tileSize;
  player.roomHeight = rooms[player.room].grid.length * config.tileSize;
  player.light = 1.0;
  sockets[player.id].emit("newRoom", player, rooms[player.room].grid, campfires[player.room]);
}

function movePlayers(){
  for(var i=0;i<players.length;i++){
    if(players[i].dead) continue;

    //decay player light
    if(players[i].light > 0) players[i].light -= plightDecay;
    if(players[i].light < 0) players[i].light = 0;

    //move player
    var pRoom = players[i].room;
    if(players[i].move.up) players[i].y -= 5;
    if(players[i].move.down) players[i].y += 5;
    if(players[i].move.left) players[i].x -= 5;
    if(players[i].move.right) players[i].x += 5;

    players[i].x = clamp(players[i].x, config.playerSize/2, players[i].roomWidth-config.playerSize/2-1);
    players[i].y = clamp(players[i].y, config.playerSize/2, players[i].roomHeight-config.playerSize/2-1);

    if(players[i].xoffset < players[i].x - players[i].sWidth + config.border){ //player too much to the right
      players[i].xoffset += (players[i].x - players[i].sWidth + config.border - players[i].xoffset) * speed;
    }
    else if(players[i].xoffset > players[i].x - config.border){ //player too much to the left
      players[i].xoffset += (players[i].x - config.border - players[i].xoffset) * speed;
    }

    if(players[i].yoffset < players[i].y - players[i].sHeight + config.border){ //player too much down
      players[i].yoffset += (players[i].y - players[i].sHeight + config.border - players[i].yoffset) * speed;
    }
    else if(players[i].yoffset > players[i].y - config.border){ //player too much up
      players[i].yoffset += (players[i].y - config.border - players[i].yoffset) * speed;
    }

    players[i].xoffset = clamp(players[i].xoffset, 0, players[i].roomWidth-players[i].sWidth-1);
    players[i].yoffset = clamp(players[i].yoffset, 0, players[i].roomHeight-players[i].sHeight-1);

    var top = (players[i].y-config.playerSize/2)/config.tileSize >> 0;
    var bottom = (players[i].y+config.playerSize/2)/config.tileSize >> 0;
    var left = (players[i].x-config.playerSize/2)/config.tileSize >> 0;
    var right = (players[i].x+config.playerSize/2)/config.tileSize >> 0;

    if(rooms[pRoom].grid[top][left] === "1" || rooms[pRoom].grid[top][right] === "1" || rooms[pRoom].grid[bottom][left] === "1" || rooms[pRoom].grid[bottom][right] === "1"){
      players[i].dead = true;
      sockets[players[i].id].emit("dead", players[i]);
      newLight(players[i].x, players[i].y, pRoom);

      var index = findID(players[i].id);
      if(index > -1) players.splice(index, 1);
    }

    for(var j=0;j<campfires[pRoom].length;j++){ //refresh player light
      if((left === campfires[pRoom][j].x || right === campfires[pRoom][j].x) &&  (top === campfires[pRoom][j].y || bottom === campfires[pRoom][j].y)){
        players[i].light = 1.0;
      }
    }

    if(left >= rooms[pRoom].finishl && right <= rooms[pRoom].finishr && top >= rooms[pRoom].finishu && bottom <= rooms[pRoom].finishd){ //finish level
      loadNextLevel(players[i]);
    }
  }
}

function update(){
  players.forEach(function(p){
    if(p.dead) return;

    var playersSection = [];
    players.forEach(function(pp){
      if(p.id !== pp.id &&
        pp.x+config.playerSize/2 > p.xoffset &&
        pp.x-config.playerSize/2 < p.xoffset+p.sWidth &&
        pp.y+config.playerSize/2 > p.yoffset &&
        pp.y-config.playerSize/2 < p.yoffset+p.sHeight &&
        p.room === pp.room) playersSection.push({x: pp.x, y: pp.y, hue: pp.hue, light: pp.light});
    });
    var player = {x: p.x, y: p.y, xoffset: p.xoffset, yoffset: p.yoffset, light: p.light};
    sockets[p.id].emit("newPosition", player, playersSection, players.length);

    var lightsSection = [];
    for(var j=0;j<lights[p.room].length;j++){
      if(lights[p.room][j].x+config.radius > p.xoffset &&
         lights[p.room][j].x-config.radius < p.xoffset+p.sWidth &&
         lights[p.room][j].y+config.radius > p.yoffset &&
         lights[p.room][j].y-config.radius < p.yoffset+p.sHeight) lightsSection.push(lights[p.room][j]);
    }
    sockets[p.id].emit("newLights", lightsSection);
  });
}

setInterval(movePlayers, 16);
setInterval(update, 16);
